Extending annotations by adding units

magdalenafuentes opened the pull request:
Following the discussion in #58 I think we should follow ideas in `mirdata` and add `units` to the annotations, so they become "reusable", e.g. we use the same annotation type for annotations with `azimuth` in `degrees` or `radians` by simply indicating the `units`.

Note that I introduced the `units` some default values, in particular for `confidence_unit="binary"`, `labels_unit="open"` and `intervals_unit="s"`. I think this is correct for all datasets that we have now but please **keep this in mind when reviewing the code because it could introduce silent bugs**. If you prefer I can make the units mandatory but that means having to modify all tests and loaders.

magdalenafuentes committed:
Add units to Annotations

magdalenafuentes committed:
Add tests

magdalenafuentes committed:
Cover missing lines tests

magdalenafuentes commented:
@iranroman once we merge this you can start with #58

justinsalamon commented on soundata/annotations.py:
```diff
@@ -2,6 +2,21 @@
+}
+
+#: Label units
+LABEL_UNITS = {"open": "no scrict schema or units"}
```
strict :)

justinsalamon commented on soundata/annotations.py:
```diff
@@ -2,6 +2,21 @@
+#: Confidence units
+CONFIDENCE_UNITS = {
+    "likelihood": "score between 0 and 1",
+    "binary": "0 or 1",
```
Do we need to have both binary and likelihood? can't we have just likelihood where datasets that are sure about events use a confidence value of 1?

justinsalamon replied:
thinking about it more, if we only allow confidence to be a float between 0-1, then we don't even need to explicitly store the unit. What do you think?

magdalenafuentes replied:
Is it worth to have two different "units" depending if the confidence is derived from the estimation of a model or manually labeled?

justinsalamon replied:
I'm not sure - a human annotator can use 1.0 to indicate they are confident in their annotations. Practically there's no different in how the values would be used versus a binary confidence type. Indicating how the dataset was annotated is more of a metadata thing, right? I think it could potentially be beneficial to enforce a consistent "model" for confidence (float in [0,1]) for both humans and machines.

Perhaps we can start without a unit for confidence (keeping it always as a float in [0, 1]), and only add it if we see there is a real need for it based on user feedback?

justinsalamon commented on soundata/annotations.py:
```diff
@@ -2,6 +2,21 @@
+#: Time units
+TIME_UNITS = {
+    "s": "seconds",
+    "ms": "miliseconds",
```
Should we use abbreviations, or should we use `seconds` and `milliseconds` directly as the value stored in the annotation? Wouldn't that be clearer/more explicit?

justinsalamon commented on soundata/annotations.py:
```diff
@@ -36,21 +62,40 @@ class Events(Annotation):
+        confidence=None,
+        intervals_unit="s",
+        labels_unit="open",
+        confidence_unit="binary",
```
I worry that if these values have defaults people will miss/ignore them and that will leads to bugs (e.g. dataset annotated in milliseconds but the interval unit is set to `s`. 

@magdalenafuentes thoughts about not having default values here and forcing the user to set them explicitly?

magdalenafuentes replied:
Yeah I agree.. If we enforce it then we have to edit all the loaders. I'm ok doing that because we don't have as many loaders yet, but in the future it would be nice to think of a way of introducing changes without having to manually edit all files. I haven't really found a workaround for it..

justinsalamon replied:
Yes, you raise a good point (about updating loaders). But hopefully the API (especially for mandatory fields) doesn't change very often (changing mandatory fields would be a breaking change so ideally we avoid it as much as possible), so I don't anticipate us having to make these updates very often at least.

If you're OK with it, I would vote for making these mandatory.

magdalenafuentes replied:
I agree on making them mandatory. About the breaking change, I super agree, but the changes I'm talking about are not always breaking changes, we can break things under the hood without the user noting and those usually need an update of loaders, which is not ideal. Anyway, just dropping that idea here so we keep it in mind for the future, let's see if we can come up with a strategy to avoid those

justinsalamon commented on soundata/annotations.py:
```diff
@@ -145,7 +190,7 @@ def validate_lengths_equal(array_list):
 
 
-def validate_confidence(confidence):
+def validate_confidence(confidence, confidence_unit="binary"):
```
If we keep confidence units as a thing, then should this have a default value? can't that lead to bugs? Isn't it better to force the user to set it?

justinsalamon commented on tests/test_annotations.py:
```diff
@@ -37,12 +40,22 @@ def test_tags():
     intervals = np.array([[1.0, 2.0], [1.5, 3.0], [2.0, 3.0]])
     labels = ["Siren", "Laughter", "Engine"]
-    confidence = np.array([1, 0.5, 0.2])
+    confidence = np.array([1.0, 0.0, 1.0])
```
just checking - do we have other tests that use confidence values other than 1/0 after this change?

justinsalamon commented on tests/test_annotations.py:
```diff
@@ -145,3 +173,15 @@ def test_validate_intervals():
+        annotations.validate_unit("c", {"a": "asdf", "b": "asdfd"})
+
+    with pytest.raises(ValueError):
+        annotations.validate_unit(None, {"a": "asdf", "b": "asdfd"})
```
Maybe use less random strings here just for clarity of the test? :)

justinsalamon reviewed (commented):
@magdalenafuentes nice work! Added some comments, happy to discuss to iron out quickly. tnx!

magdalenafuentes committed:
Address comments and make units mandatory

magdalenafuentes committed:
Update loaders

magdalenafuentes committed:
Merge branch 'master' into fuentes/annotations_extensions

magdalenafuentes committed:
Proposal to change MultiAnnotation [breaking change]

magdalenafuentes committed:
Merge branch 'fuentes/annotations_extensions' of github.com:soundata/soundata into fuentes/annotations_extensions

magdalenafuentes committed:
Update fsdnoisy18k.py

magdalenafuentes committed:
black

magdalenafuentes commented:
@justinsalamon I made the changes we discussed, let me know how it looks. Also, I added a commit with a proposal to change the `MultiAnnotator` class: I changed `labels` -> `annotations` to be more general, so now the class has the fields `annotations` and `annotators`

justinsalamon commented on soundata/annotations.py:
```diff
@@ -17,15 +26,21 @@ class Tags(Annotation):
 
     Attributes:
         labels (list): list of string tags
+        confidence (np.ndarray or None): array of confidence values
```
we can add: "...values, float in [0,1]"

justinsalamon commented on soundata/annotations.py:
```diff
@@ -36,20 +51,35 @@ class Events(Annotation):
         labels (list): list of event labels (as strings)
-        confidence (np.ndarray or None): array of float confidence values
-            in the range [0, 1]
+        confidence (np.ndarray or None): array of confidence values
```
you don't want to mention the range [0,1]?

justinsalamon commented on soundata/annotations.py:
```diff
@@ -36,20 +51,35 @@ class Events(Annotation):
-        confidence (np.ndarray or None): array of float confidence values
-            in the range [0, 1]
+        confidence (np.ndarray or None): array of confidence values
+        labels_unit (str): labels unit, one of LABELS_UNITS
```
labels unit (str)

justinsalamon replied:
ahh sorry scratch that! it's already there 🌴

justinsalamon commented on soundata/annotations.py:
```diff
@@ -59,16 +89,16 @@ class MultiAnnotator(Annotation):
 
         self.annotators = annotators
-        self.labels = labels
+        self.annotations = annotations
```
good call!

this is starting to look more and more like JAMS 😬

justinsalamon reviewed (approved):
@magdalenafuentes looks awesome. mentioned a couple of tiny nits. 🚀

justinsalamon commented:
Be sure to bump the version after you merge and make a new release :)

magdalenafuentes committed:
Complete readme annotations

magdalenafuentes committed:
Bump version
